### PR TITLE
ENT-3328: 1898563: move syspurpose subcommands within the 'syspurpose' command

### PR DESCRIPTION
**Pull request by ptoscano**
As an effort to consolidate the handling of all the syspurpose attributes in a single place, make sure that all the syspurpose attributes are handled as subcommands of the `syspurpose` command (leaving the old commands as deprecated).

First of all, drop the command name from `args` when parsing them: according to the documentation, `ArgumentParser.parse_known_args()` seems to work without the command name as first argument. Since `CliCommand` is parsing the parameters of a specific command, we can drop the command itself in addition to the application name before calling `ArgumentParser.parse_known_args()`; this avoids the need to drop it after parsing. In addition, tweak the check for missing test arguments: `if not foo` is also matched by an empty list, which now is a valid list of arguments to pass for tests. Hence, change it to check for `None` explicitly. Adapt the tests accordingly.

Second, bind the `ArgumentParser` option result of the `--show` parameter of the `syspurpose` command to a less generic variable name to avoid clashes: by default, a `--show` option in `ArgumentParser` is saved as a namespace variable `show`, which will conflicts with the various `--show` options of the subparsers of the other syspurpose subcommands.

Make `AbstractSyspurposeCommand` able to override the `ArgumentParser` created by `AbstractCLICommand` (its grandparent class), so it is possible to create a subparser instead of a standalone parser.

Then, make `SyspurposeCommand` create all the syspurpose subcommands, with their parsers created as subparsers of it; properly dispatch the parsed command line options to the right subcommand when `SyspurposeCommand` is executed.

Finally, change the short description of each subcommand to say that is deprecated when shown in the list of the subscription-command commands.

### CODE REVIEW

**Review by tlhmerry9 — COMMENTED** [review]
Besides questioning the term deprectated (I have heard that but also depreciated).  Everything else looks ok.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Hi, thanks for PR. I miss here several things:

* Update of bash completion scripts
* Update of man pages
* Unit tests for new syspurpose commands

I did quick testing of `subscription-manager show role --set "SP Starter"` and it doesn't set value for some reason. ``subscription-manager role --set "SP Starter"` works as expected. More details here:

```
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --list
+-------------------------------------------+
               Available role
+-------------------------------------------+
 - SP Server
 - SP Starter
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --show
Current Role: sp server
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --set "SP Starter"
role set to "SP Starter".
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --show
Current Role: sp server
[root@thinkpad-t580 subscription-manager]# cat /etc/rhsm/syspurpose/syspurpose.json 
{
  "foo": [
    "FOO1"
  ],
  "role": "sp server",
  "usage": "production"
}
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager role --set "SP Starter"
role set to "SP Starter".
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --show
Current Role: SP Starter
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --unset
role unset.
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --show
Current Role: SP Starter
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --unset
role unset.
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --show
Current Role: SP Starter
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager role --unset
role unset.
[root@thinkpad-t580 subscription-manager]# PYTHONPATH=./src python -m subscription_manager.scripts.subscription_manager syspurpose role --show
Role not set.
```

File: src/subscription_manager/cli_command/abstract_syspurpose.py [hunk 1]
```diff
@@ -35,13 +35,19 @@
                         "by the entitlement server administrator.\n{advice}")
 SP_ADVICE = _("If you'd like to overwrite the server side change please run: {command}")
 
+# TRANSLATORS: this refers to a deprecated command
+DEPRECATED_COMMAND_MESSAGE = _("Deprecated, see 'syspurpose'")
```
[the remark this answers is not in the document]
Comment (ptoscano) [reply src/subscription_manager/cli_command/abstract_syspurpose.py:39]: Yes, I think "deprecated" is the right word in the context of software.

File: src/subscription_manager/cli_command/abstract_syspurpose.py [hunk 2]
```diff
@@ -35,13 +35,19 @@
...
+
 
 class AbstractSyspurposeCommand(CliCommand):
     """
     Abstract command for manipulating an attribute of system purpose.
     """
 
-    def __init__(self, name, shortdesc=None, primary=False, attr=None, commands=('set', 'unset', 'show', 'list')):
+    def __init__(self, name, subparser, shortdesc=None, primary=False, attr=None, commands=('set', 'unset', 'show', 'list')):
```
[the remark this answers is not in the document]
Comment (ptoscano) [reply src/subscription_manager/cli_command/abstract_syspurpose.py:47]: I thought about it, and my idea was to force subclasses to provide a subparser. This way it is more explicit in case we add new syspurpose subcommands, which I guess we want only as subcommands of `syspurpose` and no more as top-level commands.

File: test/cli_command_test/test_config.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def test_list_default(self):
         config_command.conf['rhsmd']['processtimeout'] = '300'
         with Capture() as cap:
             self.cc._do_command = self._orig_do_command
-            self.cc.main([""])
+            self.cc.main([])
```
[the remark this answers is not in the document]
Comment (ptoscano) [reply test/cli_command_test/test_config.py:39]: This is related to the changes to `args` done in `CliCommand.main(). See the message of the first commit of this PR, which provides a lengthy (I hope) explanation of this.

File: src/subscription_manager/cli_command/cli.py [hunk 4]
```diff
@@ -236,13 +236,11 @@ def main(self, args=None):
         config_changed = False
 
         # In testing we sometimes specify args, otherwise use the default:
-        if not args:
-            args = sys.argv[1:]
+        if args is None:
+            args = sys.argv[2:]
```
[the remark this answers is not in the document]
Comment (ptoscano) [reply src/subscription_manager/cli_command/cli.py:240]: See message of the first commit.

> BTW: the comment does not make sense now.

Do you mean
```python
        # In testing we sometimes specify args, otherwise use the default:
```
? This is still valid: there are few tests that call `.main()` without args; not sure whether that makes sense or was intended, rather than passing an empty list of arguments.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Thanks for the update, but I'm still not able to set system purpose values using new commands.

File: src/subscription_manager/cli_command/cli.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ptoscano) [reply src/subscription_manager/cli_command/cli.py:240]: I amended (few days ago, forgot to mention it here) the PR text to include also these changes; would it be acceptable?

**Comment by ptoscano** [comment]
>     * Update of bash completion scripts
>     * Update of man pages

Done.

>     * Unit tests for new syspurpose commands

Still TBD, getting there.

> I did quick testing of `subscription-manager show role --set "SP Starter"` and it doesn't set value for some reason. ``subscription-manager role --set "SP Starter"` works as expected.

I tried to reproduce it (trying at different times in the past few days), unfortunately without success :-( For me, both the old and the new commands generally work (one exception below), and I see `/etc/rhsm/syspurpose/syspurpose.json` correctly updated.

The `service-level` command still doesn't fully work as syspurpose subcommand: this happens because it has its own implementation on top of `AbstractSyspurposeCommand`.

In the meanwhile, I fixed a bug that affected the `syspurpose --show` vs `syspurpose $subcommand --show` clash.

**Comment by jirihnidek** [comment]
I can confirm that it works as expected now. I will look at this PR properly tomorrow morning.

File: man/subscription-manager.8 [hunk 5]
```diff
@@ -388,15 +376,34 @@ command displays the current configured syspurpose
 .I preferences
 for the system.
 
+.PP
+The
+.B syspurpose
+command has subcommands for all the various syspurpose preferences and attributes:
```
Comment (jirihnidek) [inline man/subscription-manager.8:382]: I can remember that we had to change terminology "subcommand" in the past, because this terminology is wrong. I would write something like this: "command has next level of commands for all the various system purpose preferences and attributes:"

File: man/subscription-manager.8 [hunk 6]
```diff
@@ -388,15 +376,34 @@ command displays the current configured syspurpose
...
 Shows the system's current set of syspurpose preference formatted as JSON. Single-valued entries for which there is no value will be included in the output with a value of "". List entries which have no value will be included in the output with a value of "[]" (less the quotes).
 
 
-.SS ADDONS OPTIONS
+.PP
+.SS addons options
 The
 .B addons
-command displays the current configured addons system purpose attribute
+subcommand displays the current configured addons system purpose attribute
```
Comment (jirihnidek) [inline man/subscription-manager.8:406]: I would just avoid of using word "subcommand" and I would use just command here.

**Review by jirihnidek — CHANGES_REQUESTED** [review]
Thanks for PR. The code looks OK. :+1: I have only few requests to terminology in man pages. I can also confirm that bash completion works as expected.

**Review by jirihnidek — APPROVED** [review]
ACK, the term "subcommand" is OK according our discussion.
